CBlockIndex/CDiskBlockIndex improvements for safety, consistent behavior

jonatack opened the pull request:
Fix a few design issues, potential footguns and inconsistent behavior in the CBlockIndex and CDiskBlockIndex classes.

- Ensure phashBlock in `CBlockIndex#GetBlockHash()` is not nullptr before dereferencing and remove a now-redundant assert preceding a GetBlockHash() caller.  This protects against UB here, and in case of failure (which would indicate a consensus bug), the debug log will print `bitcoind: chain.h:265: uint256 CBlockIndex::GetBlockHash() const: Assertion 'phashBlock != nullptr' failed. Aborted` instead of `Segmentation fault`.
- Remove the unused `CDiskBlockIndex#ToString()` class member, and mark the inherited `CBlockIndex#ToString()` public interface member as deleted to disallow calling it in the derived CDiskBlockIndex class.
- Rename the `CDiskBlockIndex GetBlockHash()` class member to `ConstructBlockHash()`, which also makes sense as they perform different operations to return a blockhash, and mark the inherited `CBlockIndex#GetBlockHash()` public interface member as deleted to disallow calling it in the derived CDiskBlockIndex class.
- Move `CBlockIndex#ToString()` from header to implementation, which also allows dropping `tinyformat.h` from the header file.

Rationale and discussion regarding the CDiskBlockIndex changes:

Here is a failing test on master that demonstrates the inconsistent behavior of the current design: calling the same inherited public interface functions on the same CDiskBlockIndex object should yield identical behavior, but does not.

```diff
diff --git a/src/test/validation_chainstatemanager_tests.cpp b/src/test/validation_chainstatemanager_tests.cpp
index 6dc522b421..dac3840f32 100644
--- a/src/test/validation_chainstatemanager_tests.cpp
+++ b/src/test/validation_chainstatemanager_tests.cpp
@@ -240,6 +240,15 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_activate_snapshot, TestChain100Setup)

     const CBlockIndex* tip = chainman.ActiveTip();

     BOOST_CHECK_EQUAL(tip->nChainTx, au_data.nChainTx);

+    // CDiskBlockIndex "is a" CBlockIndex, as it publicly inherits from it.
+    // Test that calling the same inherited interface functions on the same
+    // object yields identical behavior.
+    CDiskBlockIndex index{tip};
+    CBlockIndex *pB = &index;
+    CDiskBlockIndex *pD = &index;
+    BOOST_CHECK_EQUAL(pB->GetBlockHash(), pD->GetBlockHash());
+    BOOST_CHECK_EQUAL(pB->ToString(), pD->ToString());
```

(build and run: `$ ./src/test/test_bitcoin -t validation_chainstatemanager_tests`)

The GetBlockHash() test assertion only passes on master because the different methods invoked by the current design happen to return the same result.  If one of the two is changed, it fails like the ToString() assertion does.

Redefining inherited non-virtual functions is well-documented as incorrect design to avoid inconsistent behavior (see Scott Meyers, Effective C++, Item 36). Class usage is confusing when the behavior depends on the pointer definition instead of the object definition (static binding happening where dynamic binding was expected). This can lead to unsuspected or hard-to-track bugs.

Outside of critical hot spots, correctness usually comes before optimisation, but the current design dates back to main.cpp and it may possibly have been chosen to avoid the overhead of dynamic dispatch.  This solution does the same: the class sizes are unchanged and no vptr or vtbl is added.

There are better designs for doing this that use composition instead of inheritance, or that separate the public interface from the private implementations.  One example of the latter would be a non-virtual public interface that calls private virtual implementation methods, i.e. the Template pattern via the Non-Virtual Interface (NVI) idiom.

jonatack committed:
CDiskBlockIndex: rename GetBlockHash() to ConstructBlockHash()

and mark the inherited CBlockIndex#GetBlockHash public interface member
as deleted, to disallow calling it in the derived CDiskBlockIndex class.

Here is a failing test on master demonstrating the inconsistent behavior of the
current design: calling the same inherited public interface functions on the
same CDiskBlockIndex object should yield identical behavior.

```diff
diff --git a/src/test/validation_chainstatemanager_tests.cpp b/src/test/validation_chainstatemanager_tests.cpp
index 6dc522b421..dac3840f32 100644
--- a/src/test/validation_chainstatemanager_tests.cpp
+++ b/src/test/validation_chainstatemanager_tests.cpp
@@ -240,6 +240,15 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_activate_snapshot, TestChain100Setup)

     const CBlockIndex* tip = chainman.ActiveTip();

     BOOST_CHECK_EQUAL(tip->nChainTx, au_data.nChainTx);

+    // CDiskBlockIndex "is a" CBlockIndex, as it publicly inherits from it.
+    // Test that calling the same inherited interface functions on the same
+    // object yields identical behavior.
+    CDiskBlockIndex index{tip};
+    CBlockIndex *pB = &index;
+    CDiskBlockIndex *pD = &index;
+    BOOST_CHECK_EQUAL(pB->GetBlockHash(), pD->GetBlockHash());
+    BOOST_CHECK_EQUAL(pB->ToString(), pD->ToString());
+
```

The GetBlockHash() test assertion only passes on master because the different
methods invoked by the current design happen to return the same result.  If one
of the two is changed, it fails like the ToString() assertion does.

Redefining inherited non-virtual functions is well-documented as incorrect
design to avoid inconsistent behavior (see Scott Meyers, "Effective C++", Item
36).  Class usage is confusing when the behavior depends on the pointer
definition instead of the object definition (static binding happening where
dynamic binding was expected).  This can lead to unsuspected or hard-to-track
bugs.

Outside of critical hot spots, correctness usually comes before optimisation,
but the current design dates back to main.cpp and it may possibly have been
chosen to avoid the overhead of dynamic dispatch.  This solution does the same:
the class sizes are unchanged and no vptr or vtbl is added.

There are better designs for doing this that use composition instead of
inheritance or that separate the public interface from the private
implementations.  One example of the latter would be a non-virtual public
interface that calls private virtual implementation methods, i.e. the Template
pattern via the Non-Virtual Interface (NVI) idiom.

jonatack committed:
CDiskBlockIndex: remove unused ToString() class member

and mark its inherited CBlockIndex#ToString public interface member
as deleted, to disallow calling it in the derived CDiskBlockIndex class.

jonatack commented:
Re-pushed to fix a commit message.

laanwj commented on src/chain.h:
```diff
@@ -414,16 +414,9 @@ class CDiskBlockIndex : public CBlockIndex
     }
 
-
-    std::string ToString() const
```
I'm somewhat ambivalent about removing this. The superclass has it so it can be considered part of the API, and it can be useful to show the contents of the structure when debugging/troubleshooting.

jonatack replied:
Should we rename it instead?

(Noting that the safe and correct way would be to use dynamic dispatch, like in https://github.com/jonatack/bitcoin/commits/CBlockIndex-CDiskBlockIndex-class-design, if the base and derived classes aren't critical hotspots.)

laanwj replied:
I agree it's not worth it to set that up just for a debugging function.

jonatack replied:
Right.

jonatack replied:
Resolving as the method is unused and can have inconsistent behavior as described in the OP, but please let me know if a change should be made.

jonatack commented:
The same approach and a slightly different one were also suggested in https://github.com/bitcoin/bitcoin/pull/25311#discussion_r895318403.

vasild reviewed (approved):
ACK 905a4d1b2c9ed6b28f2bfbb156bbb95b1d99c44e

Removing unused code is always good IMO. If needed it can be resurrected from git history or re-implemented.

Some non-blocker questions:

Why is overriding non-virtual functions incorrect design?

I think inheritance implies that the derived class wants to have all the properties of the base class. If it wants just some of them, then at some point maybe [composition would be more appropriate than inheritance](https://google.github.io/styleguide/cppguide.html#Inheritance).

The test from the commit message is expected to fail. If you access a Derived object through a pointer to Base, and both have the same method, then the Base's method is to be called.

Edit: now I see that the alternative approach in https://github.com/bitcoin/bitcoin/pull/25311#discussion_r895318403 uses composition instead of inheritance. If indeed the users of `CDiskBlockIndex` do not use all the stuff provided by `CBlockIndex`, then that is a better approach.

MarcoFalke commented on src/chain.h:
```diff
@@ -414,16 +414,9 @@ class CDiskBlockIndex : public CBlockIndex
-    }
+    // To be called only by the base class.
+    uint256 GetBlockHash() = delete;
+    uint256 ToString() = delete;
```
Deleted functions may also return `void`

vasild replied:
It may return anything because it is irrelevant, but if it is going to be different than the return type of the base method, then I would prefer it to return [`std::vector<double>`](https://purple.bikeshed.com/).

jonatack replied:
> Why return `uint256` from `ToString()`? It is irrelevant in this case, but better have the same return type as the parent `CBlockIndex::ToString()` in order to avoid people asking this question :)

Agree, fixed (thanks!)

jonatack commented:
Updated per `git range-diff 02ede4f 905a4d1 7866c85` to take review feedback (thanks!)

> Some non-blocker questions:
> 
> Why is overriding non-virtual functions incorrect design?

To avoid inconsistent behavior. Public inheritance establishes an invariant over specialization for the derived class. Class usage is confusing when the behavior depends on the pointer definition instead of the object definition (static binding happening where dynamic binding was expected). This can lead to unsuspected or hard-to-track bugs. See Scott Meyers, Effective C++, Item 36: Inherited non-virtual functions should never be redefined, rephrased in urls like https://programmersought.com/article/50007468749/ and https://www.programminghunter.com/article/96792264397/.
 
> Edit: now I see that the alternative approach in [#25311 (comment)](https://github.com/bitcoin/bitcoin/pull/25311#discussion_r895318403) uses composition instead of inheritance. If indeed the users of `CDiskBlockIndex` do not use all the stuff provided by `CBlockIndex`, then that is a better approach.

It's a more invasive refactoring, agree it could potentially be a better design as a follow-up.

vasild reviewed (approved):
ACK 7866c85387c3fa553437be1e004986b889fcd2de

Thanks for the explanation and links!

Now I have justification to clarify the naming of `CNetAddr` and `CService` methods which has been confusing me for a long time: https://github.com/bitcoin/bitcoin/pull/25619

MarcoFalke commented on src/chain.h:
```diff
@@ -414,16 +414,9 @@ class CDiskBlockIndex : public CBlockIndex
-            hashPrev.ToString());
-        return str;
-    }
+    // To be called only by the base class.
```
I am pretty sure that constructing a `CDiskBlockIndex`, then calling the base class's method on the constructed object will result in dereferencing a nullptr (UB)

vasild replied:
I don't get it. Do you mean this?

```cpp
class B
{
public:
    void f() { std::cout << "B::f()\n"; }
};

class D : public B
{
public:
    void f() = delete;
};

int main(int, char**)
{
    D d;
    static_cast<B*>(&d)->f();
    return 0;
}
```

(it prints "B::f()")

jonatack replied:
Running mainnet with the following patch:
```diff
diff --git a/src/txdb.cpp b/src/txdb.cpp
index bad3bb80a9..5078ddb6bf 100644
--- a/src/txdb.cpp
+++ b/src/txdb.cpp
@@ -310,7 +310,9 @@ bool CBlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams,
             CDiskBlockIndex diskindex;
             if (pcursor->GetValue(diskindex)) {
                 // Construct block index object
-                CBlockIndex* pindexNew = insertBlockIndex(diskindex.ConstructBlockHash());
+                CBlockIndex* pindexNew{insertBlockIndex(diskindex.ConstructBlockHash())};
+                assert(pindexNew);
+                LogPrintf("CBI 1: %s\n", pindexNew->ToString());
                 pindexNew->pprev          = insertBlockIndex(diskindex.hashPrev);
                 pindexNew->nHeight        = diskindex.nHeight;
                 pindexNew->nFile          = diskindex.nFile;
@@ -323,6 +325,7 @@ bool CBlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams,
                 pindexNew->nNonce         = diskindex.nNonce;
                 pindexNew->nStatus        = diskindex.nStatus;
                 pindexNew->nTx            = diskindex.nTx;
+                LogPrintf("CBI 2: %s\n", pindexNew->ToString());
 
                 if (!CheckProofOfWork(pindexNew->GetBlockHash(), pindexNew->nBits, consensusParams)) {
                     return error("%s: CheckProofOfWork failed: %s", __func__, pindexNew->ToString());
```
prints (many) valid-looking objects.
```
2022-07-15T16:34:08Z CBI 1: CBlockIndex(pprev=0, nHeight=0, merkle=0000000000000000000000000000000000000000000000000000000000000000, hashBlock=000000000000000000597281aa1ec7e99f0fff6eeda0db8707f77709951ba61a)
2022-07-15T16:34:08Z CBI 2: CBlockIndex(pprev=0x55d02f11c778, nHeight=478600, merkle=40e759d1ea04b507003f1177a66c39ec2783eb464f7d867c3d7f401fb168f2aa, hashBlock=000000000000000000597281aa1ec7e99f0fff6eeda0db8707f77709951ba61a)

2022-07-15T16:34:08Z CBI 1: CBlockIndex(pprev=0, nHeight=0, merkle=0000000000000000000000000000000000000000000000000000000000000000, hashBlock=00000000000000000506e18c023143c5c3421399d3582be8aa4b263b8f69a61a)
2022-07-15T16:34:08Z CBI 2: CBlockIndex(pprev=0x55d02f11c918, nHeight=427311, merkle=29af1206ec66690b1a0e51c2b07e6ead88056e7c11746c4d448f47e161cbbe39, hashBlock=00000000000000000506e18c023143c5c3421399d3582be8aa4b263b8f69a61a)
```
and with this patch calling the base function that is deleted in the derived class:
```diff
diff --git a/src/txdb.cpp b/src/txdb.cpp
index bad3bb80a9..c048c2d92a 100644
--- a/src/txdb.cpp
+++ b/src/txdb.cpp
@@ -310,7 +310,7 @@ bool CBlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams,
             CDiskBlockIndex diskindex;
             if (pcursor->GetValue(diskindex)) {
                 // Construct block index object
-                CBlockIndex* pindexNew = insertBlockIndex(diskindex.ConstructBlockHash());
+                CBlockIndex* pindexNew = insertBlockIndex(diskindex.GetBlockHash());
                 pindexNew->pprev          = insertBlockIndex(diskindex.hashPrev);
                 pindexNew->nHeight        = diskindex.nHeight;
```
I see the following build error:
```
```
txdb.cpp:313:69: error: attempt to use a deleted function
                CBlockIndex* pindexNew = insertBlockIndex(diskindex.GetBlockHash());
                                                                    ^
./chain.h:418:13: note: 'GetBlockHash' has been explicitly marked deleted here
    uint256 GetBlockHash() = delete;
            ^
1 error generated.
```
Let me know if I'm missing something.

MarcoFalke replied:
I mean that calling the base class method on a disk index object is UB:

```diff
diff --git a/src/txdb.cpp b/src/txdb.cpp
index c048c2d92a..de1384fb61 100644
--- a/src/txdb.cpp
+++ b/src/txdb.cpp
@@ -309,6 +309,7 @@ bool CBlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams,
         if (pcursor->GetKey(key) && key.first == DB_BLOCK_INDEX) {
             CDiskBlockIndex diskindex;
             if (pcursor->GetValue(diskindex)) {
+                static_cast<CBlockIndex&>(diskindex).ToString();
                 // Construct block index object
                 CBlockIndex* pindexNew = insertBlockIndex(diskindex.GetBlockHash());
                 pindexNew->pprev          = insertBlockIndex(diskindex.hashPrev);
```

```
==563291== Invalid read of size 16
==563291==    at 0x2192B0: GetBlockHash (chain.h:266)
==563291==    by 0x2192B0: CBlockIndex::ToString[abi:cxx11]() const (chain.h:309)
==563291==    by 0x3CD954: CBlockTreeDB::LoadBlockIndexGuts(Consensus::Params const&, std::function<CBlockIndex* (uint256 const&)>) (txdb.cpp:312)
==563291==    by 0x210BBF: node::BlockManager::LoadBlockIndex(Consensus::Params const&) (blockstorage.cpp:258)
==563291==    by 0x2114AD: node::BlockManager::LoadBlockIndexDB(Consensus::Params const&) (blockstorage.cpp:323)
==563291==    by 0x410C0B: ChainstateManager::LoadBlockIndex() (validation.cpp:4148)
==563291==    by 0x222EEB: node::LoadChainstate(bool, ChainstateManager&, CTxMemPool*, bool, bool, long, long, long, bool, bool, std::function<bool ()>, std::function<void ()>) (chainstate.cpp:53)
==563291==    by 0x1749A4: AppInitMain(node::NodeContext&, interfaces::BlockAndHeaderTipInfo*) (init.cpp:1456)
==563291==    by 0x159A31: AppInit (bitcoind.cpp:234)
==563291==    by 0x159A31: main (bitcoind.cpp:278)
==563291==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==563291== 
==563291== 
==563291== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==563291==  Access not within mapped region at address 0x0
==563291==    at 0x2192B0: GetBlockHash (chain.h:266)
==563291==    by 0x2192B0: CBlockIndex::ToString[abi:cxx11]() const (chain.h:309)
==563291==    by 0x3CD954: CBlockTreeDB::LoadBlockIndexGuts(Consensus::Params const&, std::function<CBlockIndex* (uint256 const&)>) (txdb.cpp:312)
==563291==    by 0x210BBF: node::BlockManager::LoadBlockIndex(Consensus::Params const&) (blockstorage.cpp:258)
==563291==    by 0x2114AD: node::BlockManager::LoadBlockIndexDB(Consensus::Params const&) (blockstorage.cpp:323)
==563291==    by 0x410C0B: ChainstateManager::LoadBlockIndex() (validation.cpp:4148)
==563291==    by 0x222EEB: node::LoadChainstate(bool, ChainstateManager&, CTxMemPool*, bool, bool, long, long, long, bool, bool, std::function<bool ()>, std::function<void ()>) (chainstate.cpp:53)
==563291==    by 0x1749A4: AppInitMain(node::NodeContext&, interfaces::BlockAndHeaderTipInfo*) (init.cpp:1456)
==563291==    by 0x159A31: AppInit (bitcoind.cpp:234)
==563291==    by 0x159A31: main (bitcoind.cpp:278)
```

So I'd say the comment should be clarified that it is wholly illegal to call any such method on this object (whether through itself or a class it inherits from).

jonatack replied:
Will update the comment and may add following belt-and-suspenders patch that fixes this case.

```diff
    std::string ToString() const
    {
        return strprintf("CBlockIndex(pprev=%p, nHeight=%d, merkle=%s, hashBlock=%s)",
            pprev, nHeight,
            hashMerkleRoot.ToString(),
-            GetBlockHash().ToString());
+            phashBlock ? GetBlockHash().ToString() : "");
    }
```

jonatack replied:
or

```diff
     std::string ToString() const
     {
+        if (!phashBlock) return "";
         return strprintf("CBlockIndex(pprev=%p, nHeight=%d, merkle=%s, hashBlock=%s)",
```

vasild replied:
> I am pretty sure that constructing a CDiskBlockIndex, then calling the base class's method on the constructed object will result in dereferencing a nullptr (UB)

I see your point now.

```cpp
CDiskBlockIndex diskindex;
// Will crash because `phashBlock` is nullptr.
static_cast<CBlockIndex&>(diskindex).ToString();
```

But then, the following will also crash for the same reason:

```cpp
CBlockIndex index;
index.ToString();
```

The crash is not related to the inheritance or the deleted methods from this PR.

jonatack replied:
> The crash is not related to the inheritance or the deleted methods from this PR.

Yes. I'm thinking it would maybe be best to revert this PR to the commit of the last ACK and propose the CBlockIndex improvements separately.

MarcoFalke replied:
> The crash is not related to the inheritance or the deleted methods from this PR.

Correct, this is why I commented on the documentation, which seemed to encourage calling the deleted method on the base class instead.

jonatack replied:
Dropped the documentation and moved the fix that ensures phashBlock in CBlockIndex#ToString() is not nullptr before dereferencing to the first commit of this changeset. Don't hesitate to propose a code comment if you think one is still a good idea.

jonatack committed:
CBlockIndex: ensure phashBlock is not nullptr before dereferencing

and remove a now-redundant assert preceding a GetBlockHash() caller.

This protects against UB here, and in case of failure (which would
indicate a consensus bug), the debug log will print

bitcoind: chain.h:265: uint256 CBlockIndex::GetBlockHash() const: Assertion `phashBlock != nullptr' failed.
Aborted

instead of

Segmentation fault

jonatack commented:
Added a commit to ensure `phashBlock` in `CBlockIndex#ToString()` is not nullptr before dereferencing.

jonatack committed:
refactor: move CBlockIndex#ToString() from header to implementation

which allows dropping tinyformat.h from the header file.

vasild commented on src/chain.h:
```diff
@@ -263,7 +263,7 @@ class CBlockIndex
 
     uint256 GetBlockHash() const
     {
-        return *phashBlock;
+        return *Assert(phashBlock);
```
I think this should be turned into `assert(phashBlock);` (or even better `assert(phashBlock != nullptr)`). It will crash for sure even in release builds where `Assert()` is a noop. Better provide some meaningful message in that case (from the assert), before crashing. Then this assert can be removed:

https://github.com/bitcoin/bitcoin/blob/895937edb2cca3046640dc016827a216e0b6d1c1/src/validation.cpp#L2269-L2271

jonatack replied:
Done and done, thanks!

jonatack commented on src/chain.cpp:
```diff
@@ -4,13 +4,21 @@
+{
+    return strprintf("CBlockIndex(pprev=%p, nHeight=%d, merkle=%s, hashBlock=%s)",
+                     pprev, nHeight, hashMerkleRoot.ToString(),
+                     phashBlock ? GetBlockHash().ToString() : "");
```
Agree, done both, thanks!

jonatack commented:
Updated per review feedback by @vasild and @MarcoFalke (thanks!)

MarcoFalke commented on src/chain.h:
```diff
@@ -303,10 +303,10 @@ class CBlockIndex
             pprev, nHeight,
             hashMerkleRoot.ToString(),
-            GetBlockHash().ToString());
+            phashBlock ? ", hashBlock=" + phashBlock->ToString() : "");
```
0ddad240dcd43b46ee70ccfb112d66d655b5c34f:

Missing the hash is a bug. I think it is better to assert than to silently continue when there is a bug.

jonatack replied:
Good idea, updating.

MarcoFalke commented on src/chain.cpp:
```diff
@@ -4,13 +4,21 @@
 
+std::string CBlockIndex::ToString() const
+{
+    assert(phashBlock != nullptr);
```
no need to call the assert twice in this function?

vasild replied:
Right, the first commit bc7d14164b `CBlockIndex: ensure phashBlock is not nullptr before dereferencing` seems unnecessary.

jonatack replied:
Indeed. Done.

jonatack commented:
Updated per review feedback by @MarcoFalke (thanks!)

vasild reviewed (approved):
ACK 3a61fc56a0ad6ed58570350dcfd9ed2d10239b48
